### PR TITLE
Graceful Shutdown on SIGINT and Non-Blocking Services

### PR DESCRIPTION
**Pull request by rauljordan**
Hi all,

This is a PR to handle graceful shutdown of the sharding node upon signal interrupts, using similar mechanisms to Geth with respect to even including ["panic mode"](https://github.com/ethereum/go-ethereum/blob/6cf0ab38bd0af77d81aad4c104979cebee9e3e63/cmd/utils/cmd.go#L74) where a user can keep doing ctrl-c to force a shutdown.

This PR also converts the current actor services (proposers, observers, notaries) into non-blocking services via goroutines. Error handling is done within these goroutines to prevent from blocking the main sharding node thread.

## Requirements for Merge
- [x] Convert notary/proposer to non-blocking
- [x] Implement signal interrupt graceful shutdown handlers
- [x] Include Geth's panic mode of allowing for 10 signal interrupts to forcefully shut down the system

### CODE REVIEW

**Comment by terenc3t** [comment]
The flow looks good, can you check lint? it failed

File: sharding/node/backend.go [hunk 1]
```diff
@@ -104,14 +107,34 @@ func New(ctx *cli.Context) (*ShardEthereum, error) {
...
+		defer signal.Stop(sigc)
+		<-sigc
+		log.Info("Got interrupt, shutting down...")
+		go s.Close()
+		for i := 10; i > 0; i-- {
+			<-sigc
+			if i > 1 {
+				log.Warn("Already shutting down, interrupt more to panic.", "times", i-1)
+			}
+		}
```
Comment (prestonvanloon) [inline sharding/node/backend.go]: You need to do something after exiting this for loop, like a `panic()` or similar. This happens when the user has sent 10+ SIGINT to the process.

See geth's example: https://github.com/ethereum/go-ethereum/blob/6cf0ab38bd0af77d81aad4c104979cebee9e3e63/cmd/utils/cmd.go#L83

File: sharding/notary/service.go [hunk 2]
```diff
@@ -24,22 +26,29 @@ func NewNotary(smcClient *mainchain.SMCClient, shardp2p sharding.ShardP2P, shard
...
+	return nil
+}
 
+func (n *Notary) notarizeCollations() {
 	// TODO: handle this better through goroutines. Right now, these methods
 	// are blocking.
 	if n.smcClient.DepositFlag() {
 		if err := joinNotaryPool(n.smcClient); err != nil {
-			return err
+			log.Error(fmt.Sprintf("Could not fetch current block number: %v", err))
```
Comment (prestonvanloon) [inline sharding/notary/service.go]: Does log have `log.Errorf`?

Comment (rauljordan) [reply sharding/notary/service.go]: No, unfortunately

**Comment by rauljordan** [comment]
@prestonvanloon can you take a look at my changes?

**Pull request merged**